From 30a839dd28a10b8c6c71fe3cfbe29c501537458c Mon Sep 17 00:00:00 2001
From: t <tristan75015@gmail.com>
Date: Wed, 14 Oct 2026 20:19:36 +0200
Subject: [PATCH] Bindings: displacement gradient option

C, Fortran, Python and Julia bindings of
FiniteStrainBehaviourOptions::gradient and of usesDisplacementGradient.
The C and Python tests load a Hencky behaviour with both gradients.
---
 bindings/c/include/MGIS/Behaviour/Behaviour.h | 38 +++++++++
 bindings/c/src/Behaviour.cxx                  | 46 +++++++++++
 bindings/c/tests/CMakeLists.txt               | 22 ++++++
 bindings/c/tests/DisplacementGradientTest-c.c | 76 ++++++++++++++++++
 bindings/fortran/src/mgis_behaviour.f90       | 79 +++++++++++++++++++
 bindings/julia/src/Behaviour.cxx              | 25 +++++-
 bindings/python/src/Behaviour.cxx             | 15 +++-
 bindings/python/tests/CMakeLists.txt          |  5 ++
 .../python/tests/DisplacementGradientTest.py  | 37 +++++++++
 9 files changed, 341 insertions(+), 2 deletions(-)
 create mode 100644 bindings/c/tests/DisplacementGradientTest-c.c
 create mode 100644 bindings/python/tests/DisplacementGradientTest.py

diff --git a/bindings/c/include/MGIS/Behaviour/Behaviour.h b/bindings/c/include/MGIS/Behaviour/Behaviour.h
index 52e3095..8f7bb0c 100644
--- a/bindings/c/include/MGIS/Behaviour/Behaviour.h
+++ b/bindings/c/include/MGIS/Behaviour/Behaviour.h
@@ -94,6 +94,13 @@ typedef enum {
                         * deformation gradient  */
 } mgis_bv_FiniteStrainBehavourTangentOperator;
 
+//! gradients given to finite strain behaviours
+typedef enum {
+  MGIS_BV_DEFORMATION_GRADIENT = 0, /*!< \brief deformation gradient F */
+  MGIS_BV_DISPLACEMENT_GRADIENT = 1 /*!< \brief displacement gradient
+                                     *   H = F - I */
+} mgis_bv_FiniteStrainBehaviourGradient;
+
 /*!
  * \brief allocate a new
  * `mgis_bv_FiniteStrainBehaviourOptions` object
@@ -144,6 +151,28 @@ mgis_bv_finite_strain_behaviour_options_set_tangent_operator(
 MGIS_C_EXPORT mgis_status
 mgis_bv_finite_strain_behaviour_options_set_tangent_operator_by_string(
     mgis_bv_FiniteStrainBehaviourOptions* const, const char* const);
+/*!
+ * \brief set the gradient given to the behaviour: the deformation gradient
+ * (by default), or the displacement gradient, which keeps the relative
+ * precision of small strains in single precision (see the documentation of
+ * `mgis::behaviour::FiniteStrainBehaviourOptions::gradient`)
+ * \param[in] o: finite strain behaviour options
+ * \param[in] g: gradient
+ */
+MGIS_C_EXPORT mgis_status mgis_bv_finite_strain_behaviour_options_set_gradient(
+    mgis_bv_FiniteStrainBehaviourOptions* const,
+    const mgis_bv_FiniteStrainBehaviourGradient);
+/*!
+ * \brief set the gradient given to the behaviour
+ * \param[in] o: finite strain behaviour options
+ * \param[in] g: gradient
+ * g must be have the following values:
+ * - `DeformationGradient` (or `F`)
+ * - `DisplacementGradient` (or `H`)
+ */
+MGIS_C_EXPORT mgis_status
+mgis_bv_finite_strain_behaviour_options_set_gradient_by_string(
+    mgis_bv_FiniteStrainBehaviourOptions* const, const char* const);
 /*!
  * \brief free the ressources associated with a
  * `mgis_bv_FiniteStrainBehaviourOptions` object
@@ -205,6 +234,15 @@ MGIS_C_EXPORT mgis_status mgis_bv_load_finite_strain_behaviour(
  */
 MGIS_C_EXPORT mgis_status mgis_bv_behaviour_get_tangent_operator_array_size(
     mgis_size_type* const, const mgis_bv_Behaviour* const);
+/*!
+ * \brief state if the gradients of the behaviour are the displacement
+ * gradient rather than the deformation gradient (finite strain behaviour
+ * loaded with the `MGIS_BV_DISPLACEMENT_GRADIENT` option)
+ * \param[out] r: result
+ * \param[in] b: behaviour
+ */
+MGIS_C_EXPORT mgis_status mgis_bv_behaviour_uses_displacement_gradient(
+    int* const, const mgis_bv_Behaviour* const);
 /*
  * The following rotation functions report a failure for a behaviour which
  * provides no rotation function of an integration point: an isotropic
diff --git a/bindings/c/src/Behaviour.cxx b/bindings/c/src/Behaviour.cxx
index 822ea87..9cb0324 100644
--- a/bindings/c/src/Behaviour.cxx
+++ b/bindings/c/src/Behaviour.cxx
@@ -117,6 +117,42 @@ mgis_bv_finite_strain_behaviour_options_set_tangent_operator_by_string(
 }  // end of
    // mgis_bv_finite_strain_behaviour_options_set_tangent_operator_by_string
 
+mgis_status mgis_bv_finite_strain_behaviour_options_set_gradient(
+    mgis_bv_FiniteStrainBehaviourOptions* const o,
+    const mgis_bv_FiniteStrainBehaviourGradient g) {
+  if (o == nullptr) {
+    return mgis_report_failure("invalid argument");
+  }
+  if (g == MGIS_BV_DEFORMATION_GRADIENT) {
+    o->gradient =
+        mgis::behaviour::FiniteStrainBehaviourOptions::DEFORMATION_GRADIENT;
+  } else if (g == MGIS_BV_DISPLACEMENT_GRADIENT) {
+    o->gradient =
+        mgis::behaviour::FiniteStrainBehaviourOptions::DISPLACEMENT_GRADIENT;
+  } else {
+    return mgis_report_failure("invalid gradient");
+  }
+  return mgis_report_success();
+}  // end of mgis_bv_finite_strain_behaviour_options_set_gradient
+
+mgis_status mgis_bv_finite_strain_behaviour_options_set_gradient_by_string(
+    mgis_bv_FiniteStrainBehaviourOptions* const o, const char* const g) {
+  if ((o == nullptr) || (g == nullptr)) {
+    return mgis_report_failure("invalid argument");
+  }
+  if ((strcmp(g, "DeformationGradient") == 0) || (strcmp(g, "F") == 0)) {
+    o->gradient =
+        mgis::behaviour::FiniteStrainBehaviourOptions::DEFORMATION_GRADIENT;
+  } else if ((strcmp(g, "DisplacementGradient") == 0) ||
+             (strcmp(g, "H") == 0)) {
+    o->gradient =
+        mgis::behaviour::FiniteStrainBehaviourOptions::DISPLACEMENT_GRADIENT;
+  } else {
+    return mgis_report_failure("invalid gradient");
+  }
+  return mgis_report_success();
+}  // end of mgis_bv_finite_strain_behaviour_options_set_gradient_by_string
+
 mgis_status mgis_bv_free_finite_strain_behaviour_options(
     mgis_bv_FiniteStrainBehaviourOptions** o) {
   try {
@@ -354,6 +390,16 @@ mgis_status mgis_bv_behaviour_get_tangent_operator_array_size(
   return mgis_report_success();
 }  // end of mgis_bv_behaviour_get_tangent_operator_array_size
 
+mgis_status mgis_bv_behaviour_uses_displacement_gradient(
+    int* const r, const mgis_bv_Behaviour* const b) {
+  *r = static_cast<int>(false);
+  if (b == nullptr) {
+    return mgis_report_failure("invalid argument");
+  }
+  *r = static_cast<int>(mgis::behaviour::usesDisplacementGradient(*b));
+  return mgis_report_success();
+}  // end of mgis_bv_behaviour_uses_displacement_gradient
+
 mgis_status mgis_bv_behaviour_get_library(const char** l,
                                           const mgis_bv_Behaviour* const b) {
   if (b == nullptr) {
diff --git a/bindings/c/tests/CMakeLists.txt b/bindings/c/tests/CMakeLists.txt
index 27b7e4b..d49ede7 100644
--- a/bindings/c/tests/CMakeLists.txt
+++ b/bindings/c/tests/CMakeLists.txt
@@ -136,6 +136,28 @@ else((CMAKE_HOST_WIN32) AND (NOT MSYS))
     PROPERTY DEPENDS BehaviourTest)
 endif((CMAKE_HOST_WIN32) AND (NOT MSYS))
 
+# finite strain behaviour loaded with the displacement gradient, if TFEL
+# supports it
+if(EXISTS "${TFEL_INCLUDE_PATH}/MFront/GenericBehaviour/DisplacementGradient.hxx")
+  add_executable(DisplacementGradientTest-c
+    EXCLUDE_FROM_ALL
+    DisplacementGradientTest-c.c)
+  target_link_libraries(DisplacementGradientTest-c
+    PRIVATE MFrontGenericInterface-c MFrontGenericInterface)
+  add_test(NAME DisplacementGradientTest-c
+    COMMAND DisplacementGradientTest-c
+    "$<TARGET_FILE:BehaviourTest>" "HenckyElasticity")
+  add_dependencies(build-tests DisplacementGradientTest-c)
+  if((CMAKE_HOST_WIN32) AND (NOT MSYS))
+    set_property(TEST DisplacementGradientTest-c
+      PROPERTY DEPENDS BehaviourTest
+      PROPERTY ENVIRONMENT "PATH=$<TARGET_FILE_DIR:MFrontGenericInterface>\;$<TARGET_FILE_DIR:MFrontGenericInterface-c>\;${MGIS_PATH_STRING}")
+  else((CMAKE_HOST_WIN32) AND (NOT MSYS))
+    set_property(TEST DisplacementGradientTest-c
+      PROPERTY DEPENDS BehaviourTest)
+  endif((CMAKE_HOST_WIN32) AND (NOT MSYS))
+endif()
+
 add_test(NAME BehaviourDataTest-c
  COMMAND BehaviourDataTest-c
  "$<TARGET_FILE:BehaviourTest>" "Gurson")
diff --git a/bindings/c/tests/DisplacementGradientTest-c.c b/bindings/c/tests/DisplacementGradientTest-c.c
new file mode 100644
index 0000000..c0ff828
--- /dev/null
+++ b/bindings/c/tests/DisplacementGradientTest-c.c
@@ -0,0 +1,76 @@
+/*!
+ * \file   DisplacementGradientTest-c.c
+ * \brief  Finite strain behaviour loaded with the displacement gradient
+ * through the C bindings
+ * \date   14/10/2026
+ * \copyright (C) Copyright Thomas Helfer 2018.
+ * Use, modification and distribution are subject
+ * to one of the following licences:
+ * - GNU Lesser General Public License (LGPL), Version 3.0. (See accompanying
+ *   file LGPL-3.0.txt)
+ * - CECILL-C,  Version 1.0 (See accompanying files
+ *   CeCILL-C_V1-en.txt and CeCILL-C_V1-fr.txt).
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include "MGIS/Behaviour/Behaviour.h"
+
+int test_status = EXIT_SUCCESS;
+
+static void check_status(const mgis_status s) {
+  if (s.exit_status != MGIS_SUCCESS) {
+    fprintf(stderr, "invalid function call: %s\n", s.msg);
+    exit(EXIT_FAILURE);
+  }
+}  // end of check_status
+
+static int check(const int b, const char* const e) {
+  if (b == 0) {
+    test_status = EXIT_FAILURE;
+    fprintf(stderr, "%s\n", e);
+  }
+  return b;
+}  // end of check
+
+/*!
+ * \brief load the behaviour with the given gradient, selected by its name
+ * if `g` is not null, and by the enumeration otherwise
+ */
+static void test(const char* const* argv,
+                 const char* const g,
+                 const mgis_bv_FiniteStrainBehaviourGradient eg,
+                 const int expected) {
+  mgis_bv_Behaviour* b;
+  mgis_bv_FiniteStrainBehaviourOptions* o;
+  check_status(mgis_bv_create_finite_strain_behaviour_options(&o));
+  if (g != NULL) {
+    check_status(
+        mgis_bv_finite_strain_behaviour_options_set_gradient_by_string(o, g));
+  } else {
+    check_status(mgis_bv_finite_strain_behaviour_options_set_gradient(o, eg));
+  }
+  check_status(mgis_bv_load_finite_strain_behaviour(&b, o, argv[1], argv[2],
+                                                    "Tridimensional"));
+  int r;
+  check_status(mgis_bv_behaviour_uses_displacement_gradient(&r, b));
+  check(r == expected, "invalid kind of gradient");
+  mgis_size_type gs_size;
+  check_status(mgis_bv_behaviour_get_gradients_size(&gs_size, b));
+  check(gs_size == 9, "invalid gradient size");
+  check_status(mgis_bv_free_finite_strain_behaviour_options(&o));
+  check_status(mgis_bv_free_behaviour(&b));
+}  // end of test
+
+int main(const int argc, const char* const* argv) {
+  if (!check(argc == 3, "expected three arguments")) {
+    return EXIT_FAILURE;
+  }
+  test(argv, NULL, MGIS_BV_DEFORMATION_GRADIENT, 0);
+  test(argv, NULL, MGIS_BV_DISPLACEMENT_GRADIENT, 1);
+  test(argv, "DeformationGradient", MGIS_BV_DEFORMATION_GRADIENT, 0);
+  test(argv, "F", MGIS_BV_DEFORMATION_GRADIENT, 0);
+  test(argv, "DisplacementGradient", MGIS_BV_DEFORMATION_GRADIENT, 1);
+  test(argv, "H", MGIS_BV_DEFORMATION_GRADIENT, 1);
+  return test_status;
+}  // end of main
diff --git a/bindings/fortran/src/mgis_behaviour.f90 b/bindings/fortran/src/mgis_behaviour.f90
index d643bb1..e7d9234 100644
--- a/bindings/fortran/src/mgis_behaviour.f90
+++ b/bindings/fortran/src/mgis_behaviour.f90
@@ -72,6 +72,10 @@ module mgis_behaviour
      enumerator :: DPK1_DF = 2
      enumerator :: DTAU_DDF = 3
   end enum
+  enum, bind(C)
+     enumerator :: DEFORMATION_GRADIENT = 0
+     enumerator :: DISPLACEMENT_GRADIENT = 1
+  end enum
   type :: FiniteStrainBehaviourOptions
     private
     type(c_ptr) :: ptr = c_null_ptr
@@ -323,6 +327,52 @@ contains
     s = fsb_opts_set_tangent_operator_by_string_wrapper(o%ptr,convert_fortran_string(to))
   end function finite_strain_behaviour_options_set_tangent_operator_by_string
   !
+  function finite_strain_behaviour_options_set_gradient(o, g) result(s)
+    use, intrinsic :: iso_c_binding, only: c_int
+    use mgis_fortran_utilities
+    use mgis, only: mgis_status, report_failure
+    implicit none
+    interface
+       function fsb_opts_set_gradient_wrapper(o,g) &
+            bind(c,name = 'mgis_bv_finite_strain_behaviour_options_set_gradient') &
+            result(r)
+         use, intrinsic :: iso_c_binding, only: c_ptr, c_int
+         use mgis, only: mgis_status
+         implicit none
+         type(c_ptr), intent(in), value :: o
+         integer(kind=c_int), intent(in),value :: g
+         type(mgis_status) :: r
+       end function fsb_opts_set_gradient_wrapper
+    end interface
+    type(FiniteStrainBehaviourOptions), intent(in) :: o
+    integer(kind=c_int), intent(in) :: g
+    type(mgis_status) :: s
+    s = fsb_opts_set_gradient_wrapper(o%ptr,g)
+  end function finite_strain_behaviour_options_set_gradient
+  !
+  function finite_strain_behaviour_options_set_gradient_by_string(o, g) result(s)
+    use, intrinsic :: iso_c_binding, only: c_int
+    use mgis_fortran_utilities
+    use mgis, only: mgis_status, report_failure
+    implicit none
+    interface
+       function fsb_opts_set_gradient_by_string_wrapper(o,g) &
+            bind(c,name = 'mgis_bv_finite_strain_behaviour_options_set_gradient_by_string') &
+            result(r)
+         use, intrinsic :: iso_c_binding, only: c_ptr, c_char
+         use mgis, only: mgis_status
+         implicit none
+         type(c_ptr), intent(in), value :: o
+         character(len=1,kind=c_char), dimension(*), intent(in) :: g
+         type(mgis_status) :: r
+       end function fsb_opts_set_gradient_by_string_wrapper
+    end interface
+    type(FiniteStrainBehaviourOptions), intent(in) :: o
+    character(len=*), intent(in) :: g
+    type(mgis_status) :: s
+    s = fsb_opts_set_gradient_by_string_wrapper(o%ptr,convert_fortran_string(g))
+  end function finite_strain_behaviour_options_set_gradient_by_string
+  !
   function free_finite_strain_behaviour_options(ptr) result(r)
     use, intrinsic :: iso_c_binding, only: c_associated
     use mgis
@@ -777,6 +827,35 @@ contains
     s = behaviour_get_tangent_operator_array_size_wrapper(ns, b%ptr)
     n = ns
   end function behaviour_get_tangent_operator_array_size
+  ! behaviour_uses_displacement_gradient
+  function behaviour_uses_displacement_gradient(r,b) result(s)
+    use, intrinsic :: iso_c_binding, only: c_int
+    use mgis_fortran_utilities
+    use mgis, only: mgis_status
+    implicit none
+    interface
+       function behaviour_uses_displacement_gradient_wrapper(r,b) &
+            bind(c,name = 'mgis_bv_behaviour_uses_displacement_gradient') &
+            result(s)
+         use, intrinsic :: iso_c_binding, only: c_int, c_ptr
+         use mgis, only: mgis_status
+         implicit none
+         integer(kind=c_int), intent(out) :: r
+         type(c_ptr), intent(in), value :: b
+         type(mgis_status) :: s
+       end function behaviour_uses_displacement_gradient_wrapper
+    end interface
+    logical, intent(out) :: r
+    type(behaviour), intent(in) :: b
+    type(mgis_status) :: s
+    integer(kind=c_int) :: ri
+    s = behaviour_uses_displacement_gradient_wrapper(ri, b%ptr)
+    if (ri .eq. 0) then
+       r = .false.
+    else
+       r = .true.
+    end if
+  end function behaviour_uses_displacement_gradient
   ! behaviour_get_library
   function behaviour_get_library(l,b) result(s)
     use mgis_fortran_utilities
diff --git a/bindings/julia/src/Behaviour.cxx b/bindings/julia/src/Behaviour.cxx
index 34425ea..c1e7cdc 100644
--- a/bindings/julia/src/Behaviour.cxx
+++ b/bindings/julia/src/Behaviour.cxx
@@ -86,6 +86,17 @@ void declareBehaviour(jlcxx::Module &m) {
   m.set_const("DPK2_DEGL", FiniteStrainBehaviourOptions::DS_DEGL);
   m.set_const("DTAU_DDF", FiniteStrainBehaviourOptions::DTAU_DDF);
   //!
+  m.add_bits<FiniteStrainBehaviourOptions::Gradient>(
+      "FiniteStrainBehaviourOptionsGradient");
+  m.set_const("DEFORMATION_GRADIENT",
+              FiniteStrainBehaviourOptions::DEFORMATION_GRADIENT);
+  m.set_const("DeformationGradient",
+              FiniteStrainBehaviourOptions::DEFORMATION_GRADIENT);
+  m.set_const("DISPLACEMENT_GRADIENT",
+              FiniteStrainBehaviourOptions::DISPLACEMENT_GRADIENT);
+  m.set_const("DisplacementGradient",
+              FiniteStrainBehaviourOptions::DISPLACEMENT_GRADIENT);
+  //!
   m.add_bits<Behaviour::Symmetry>("BehaviourSymmetry");
   m.set_const("ISOTROPIC", Behaviour::ISOTROPIC);
   m.set_const("Isotropic", Behaviour::ISOTROPIC);
@@ -112,7 +123,16 @@ void declareBehaviour(jlcxx::Module &m) {
           [](FiniteStrainBehaviourOptions &o,
              const FiniteStrainBehaviourOptions::TangentOperator &to) noexcept {
             o.tangent_operator = to;
-          });
+          })
+      .method("get_gradient",
+              [](const FiniteStrainBehaviourOptions &o) noexcept {
+                return o.gradient;
+              })
+      .method("set_gradient!",
+              [](FiniteStrainBehaviourOptions &o,
+                 const FiniteStrainBehaviourOptions::Gradient &g) noexcept {
+                o.gradient = g;
+              });
   //
   m.add_bits<Behaviour::BehaviourType>("BehaviourType");
   m.set_const("GENERALBEHAVIOUR", Behaviour::GENERALBEHAVIOUR);
@@ -173,6 +193,9 @@ void declareBehaviour(jlcxx::Module &m) {
   //
   m.method("load", load);
   m.method("load", load2);
+  m.method("uses_displacement_gradient", [](const Behaviour &b) {
+    return mgis::behaviour::usesDisplacementGradient(b);
+  });
   m.method("set_parameter!", setParameter1);
   m.method("set_integer_parameter!", setParameter2);
   m.method("set_unsigned_short_parameter!",
diff --git a/bindings/python/src/Behaviour.cxx b/bindings/python/src/Behaviour.cxx
index 9c1a78e..65d66a8 100644
--- a/bindings/python/src/Behaviour.cxx
+++ b/bindings/python/src/Behaviour.cxx
@@ -218,6 +218,13 @@ void declareBehaviour(pybind11::module_ &m) {
       .value("DPK1_DF", FiniteStrainBehaviourOptions::DPK1_DF)
       .value("DS_DEGL", FiniteStrainBehaviourOptions::DS_DEGL)
       .value("DTAU_DDF", FiniteStrainBehaviourOptions::DTAU_DDF);
+  // wrapping the FiniteStrainBehaviourOptions::Gradient enum
+  pybind11::enum_<FiniteStrainBehaviourOptions::Gradient>(
+      m, "FiniteStrainBehaviourOptionsGradient")
+      .value("DEFORMATION_GRADIENT",
+             FiniteStrainBehaviourOptions::DEFORMATION_GRADIENT)
+      .value("DISPLACEMENT_GRADIENT",
+             FiniteStrainBehaviourOptions::DISPLACEMENT_GRADIENT);
   // wrapping the FiniteStrainBehaviourOptions class
   pybind11::class_<FiniteStrainBehaviourOptions>(m,
                                                  "FiniteStrainBehaviourOptions")
@@ -227,7 +234,10 @@ void declareBehaviour(pybind11::module_ &m) {
                      "defines the stress measure")
       .def_readwrite("tangent_operator",
                      &FiniteStrainBehaviourOptions::tangent_operator,
-                     "defines the tangent operator");
+                     "defines the tangent operator")
+      .def_readwrite("gradient", &FiniteStrainBehaviourOptions::gradient,
+                     "defines the gradient given to the behaviour: the "
+                     "deformation gradient or the displacement gradient");
   // wrapping the Behaviour class
   pybind11::class_<Behaviour, BehaviourDescription>(m, "Behaviour")
       // pybind11::class_<Behaviour>("Behaviour")
@@ -259,6 +269,9 @@ void declareBehaviour(pybind11::module_ &m) {
 
   m.def("load", load_ptr);
   m.def("load", load_ptr2);
+  m.def("usesDisplacementGradient", mgis::behaviour::usesDisplacementGradient,
+        "return true if the gradients of the behaviour are the displacement "
+        "gradient");
   m.def("loadFromDatabase", Behaviour_loadFromDatabase, "name"_a,
         "hypothesis"_a, "material"_a = std::optional<std::string>{});
 
diff --git a/bindings/python/tests/CMakeLists.txt b/bindings/python/tests/CMakeLists.txt
index 9047150..301f288 100644
--- a/bindings/python/tests/CMakeLists.txt
+++ b/bindings/python/tests/CMakeLists.txt
@@ -44,4 +44,9 @@ test_python_bindings(IntegrateTest6)
 test_python_bindings(ExternalStateVariableTest)
 test_python_bindings(InitializeFunctionTest)
 test_python_bindings(PostProcessingTest)
+# finite strain behaviour loaded with the displacement gradient, if TFEL
+# supports it
+if(EXISTS "${TFEL_INCLUDE_PATH}/MFront/GenericBehaviour/DisplacementGradient.hxx")
+  test_python_bindings(DisplacementGradientTest)
+endif()
 
diff --git a/bindings/python/tests/DisplacementGradientTest.py b/bindings/python/tests/DisplacementGradientTest.py
new file mode 100644
index 0000000..ca2aa69
--- /dev/null
+++ b/bindings/python/tests/DisplacementGradientTest.py
@@ -0,0 +1,37 @@
+# -*- coding: utf-8 -*-
+
+import os
+try:
+    import unittest2 as unittest
+except ImportError:
+    import unittest
+import mgis.behaviour as mgis_bv
+
+
+class DisplacementGradientTest(unittest.TestCase):
+    def test_pass(self):
+
+        gradients = mgis_bv.FiniteStrainBehaviourOptionsGradient
+        # path to the test library
+        lib = os.environ['MGIS_TEST_BEHAVIOURS_LIBRARY']
+        h = mgis_bv.Hypothesis.Tridimensional
+        o = mgis_bv.FiniteStrainBehaviourOptions()
+        self.assertTrue(o.gradient == gradients.DEFORMATION_GRADIENT,
+                        'invalid default gradient')
+        bF = mgis_bv.load(o, lib, 'HenckyElasticity', h)
+        o.gradient = gradients.DISPLACEMENT_GRADIENT
+        bH = mgis_bv.load(o, lib, 'HenckyElasticity', h)
+        self.assertFalse(mgis_bv.usesDisplacementGradient(bF),
+                         'unexpected displacement gradient')
+        self.assertTrue(mgis_bv.usesDisplacementGradient(bH),
+                        'displacement gradient expected')
+        self.assertTrue(bF.gradients[0].name == 'DeformationGradient',
+                        'invalid gradient name')
+        self.assertTrue(bH.gradients[0].name == 'DisplacementGradient',
+                        'invalid gradient name')
+        self.assertTrue(bH.gradients[0].type == mgis_bv.VariableType.TENSOR,
+                        'invalid gradient type')
+
+
+if __name__ == '__main__':
+    unittest.main()
-- 
2.43.0

